File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/sorafs/GatewayFetchSummary.java
package org.hyperledger.iroha.android.sorafs;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.client.JsonParser;

/**
 * Immutable view of the JSON summary returned by the SoraFS orchestrator after a gateway fetch.
 */
public final class GatewayFetchSummary {

  private final String manifestIdHex;
  private final String chunkerHandle;
  private final String clientId;
  private final long chunkCount;
  private final long contentLength;
  private final long assembledBytes;
  private final List<ProviderReport> providerReports;
  private final List<ChunkReceipt> chunkReceipts;
  private final String anonymityPolicy;
  private final String anonymityStatus;
  private final String anonymityReason;
  private final long anonymitySoranetSelected;
  private final long anonymityPqSelected;
  private final long anonymityClassicalSelected;
  private final double anonymityClassicalRatio;
  private final double anonymityPqRatio;
  private final double anonymityCandidateRatio;
  private final double anonymityDeficitRatio;
  private final double anonymitySupplyDelta;
  private final boolean anonymityBrownout;
  private final boolean anonymityBrownoutEffective;
  private final boolean anonymityUsesClassical;

  private GatewayFetchSummary(
      final Builder builder,
      final List<ProviderReport> providerReports,
      final List<ChunkReceipt> chunkReceipts) {
    this.manifestIdHex = builder.manifestIdHex;
    this.chunkerHandle = builder.chunkerHandle;
    this.clientId = builder.clientId;
    this.chunkCount = builder.chunkCount;
    this.contentLength = builder.contentLength;
    this.assembledBytes = builder.assembledBytes;
    this.providerReports = providerReports;
    this.chunkReceipts = chunkReceipts;
    this.anonymityPolicy = builder.anonymityPolicy;
    this.anonymityStatus = builder.anonymityStatus;
    this.anonymityReason = builder.anonymityReason;
    this.anonymitySoranetSelected = builder.anonymitySoranetSelected;
    this.anonymityPqSelected = builder.anonymityPqSelected;
    this.anonymityClassicalSelected = builder.anonymityClassicalSelected;
    this.anonymityClassicalRatio = builder.anonymityClassicalRatio;
    this.anonymityPqRatio = builder.anonymityPqRatio;
    this.anonymityCandidateRatio = builder.anonymityCandidateRatio;
    this.anonymityDeficitRatio = builder.anonymityDeficitRatio;
    this.anonymitySupplyDelta = builder.anonymitySupplyDelta;
    this.anonymityBrownout = builder.anonymityBrownout;
    this.anonymityBrownoutEffective = builder.anonymityBrownoutEffective;
    this.anonymityUsesClassical = builder.anonymityUsesClassical;
  }

  /** Parses a JSON payload (UTF-8 bytes) into a {@link GatewayFetchSummary}. */
  public static GatewayFetchSummary fromJsonBytes(final byte[] payload) {
    Objects.requireNonNull(payload, "payload");
    if (payload.length == 0) {
      throw new SorafsStorageException("Gateway summary JSON must not be empty");
    }
    final String json = new String(payload, StandardCharsets.UTF_8);
    final Object parsed = JsonParser.parse(json);
    if (!(parsed instanceof Map)) {
      throw new SorafsStorageException("Gateway summary must be a JSON object");
    }
    @SuppressWarnings("unchecked")
    final Map<String, Object> root = (Map<String, Object>) parsed;
    return fromJsonObject(root);
  }

  private static GatewayFetchSummary fromJsonObject(final Map<String, Object> root) {
    final Builder builder = new Builder();
    builder.manifestIdHex = requireString(root, "manifest_id_hex");
    builder.chunkerHandle = requireString(root, "chunker_handle");
    builder.clientId = optionalString(root, "client_id");
    builder.chunkCount = requireLong(root, "chunk_count");
    builder.contentLength = requireLong(root, "content_length");
    builder.assembledBytes = requireLong(root, "assembled_bytes");
    builder.anonymityPolicy = requireString(root, "anonymity_policy");
    builder.anonymityStatus = requireString(root, "anonymity_status");
    builder.anonymityReason = optionalString(root, "anonymity_reason");
    builder.anonymitySoranetSelected = requireLong(root, "anonymity_soranet_selected");
    builder.anonymityPqSelected = requireLong(root, "anonymity_pq_selected");
    builder.anonymityClassicalSelected = requireLong(root, "anonymity_classical_selected");
    builder.anonymityClassicalRatio = requireDouble(root, "anonymity_classical_ratio");
    builder.anonymityPqRatio = requireDouble(root, "anonymity_pq_ratio");
    builder.anonymityCandidateRatio = requireDouble(root, "anonymity_candidate_ratio");
    builder.anonymityDeficitRatio = requireDouble(root, "anonymity_deficit_ratio");
    builder.anonymitySupplyDelta = requireDouble(root, "anonymity_supply_delta");
    builder.anonymityBrownout = requireBoolean(root, "anonymity_brownout");
    builder.anonymityBrownoutEffective = requireBoolean(root, "anonymity_brownout_effective");
    builder.anonymityUsesClassical = requireBoolean(root, "anonymity_uses_classical");

    final List<ProviderReport> providerReports =
        parseProviderReports(requireList(root, "provider_reports"));
    final List<ChunkReceipt> chunkReceipts =
        parseChunkReceipts(requireList(root, "chunk_receipts"));

    return new GatewayFetchSummary(builder, providerReports, chunkReceipts);
  }

  private static List<ProviderReport> parseProviderReports(final List<Object> items) {
    final List<ProviderReport> reports = new ArrayList<>(items.size());
    for (final Object item : items) {
      if (!(item instanceof Map)) {
        throw new SorafsStorageException("provider_reports entries must be objects");
      }
      @SuppressWarnings("unchecked")
      final Map<String, Object> map = (Map<String, Object>) item;
      reports.add(
          new ProviderReport(
              requireString(map, "provider"),
              requireLong(map, "successes"),
              requireLong(map, "failures"),
              requireBoolean(map, "disabled")));
    }
    return Collections.unmodifiableList(reports);
  }

  private static List<ChunkReceipt> parseChunkReceipts(final List<Object> items) {
    final List<ChunkReceipt> receipts = new ArrayList<>(items.size());
    for (final Object item : items) {
      if (!(item instanceof Map)) {
        throw new SorafsStorageException("chunk_receipts entries must be objects");
      }
      @SuppressWarnings("unchecked")
      final Map<String, Object> map = (Map<String, Object>) item;
      receipts.add(
          new ChunkReceipt(
              (int) requireLong(map, "chunk_index"),
              requireString(map, "provider"),
              requireLong(map, "attempts")));
    }
    return Collections.unmodifiableList(receipts);
  }

  private static String requireString(final Map<String, Object> map, final String key) {
    final Object value = map.get(key);
    if (value instanceof String str) {
      return str;
    }
    throw new SorafsStorageException("Expected string for `" + key + "`");
  }

  private static String optionalString(final Map<String, Object> map, final String key) {
    final Object value = map.get(key);
    if (value == null) {
      return null;
    }
    if (value instanceof String str) {
      return str;
    }
    throw new SorafsStorageException("Expected string for `" + key + "`");
  }

  private static long requireLong(final Map<String, Object> map, final String key) {
    final Object value = map.get(key);
    if (value instanceof Number number) {
      if (number instanceof Float || number instanceof Double) {
        throw new SorafsStorageException("Expected integer for `" + key + "`");
      }
      return number.longValue();
    }
    throw new SorafsStorageException("Expected number for `" + key + "`");
  }

  private static double requireDouble(final Map<String, Object> map, final String key) {
    final Object value = map.get(key);
    if (value instanceof Number number) {
      return number.doubleValue();
    }
    throw new SorafsStorageException("Expected number for `" + key + "`");
  }

  private static boolean requireBoolean(final Map<String, Object> map, final String key) {
    final Object value = map.get(key);
    if (value instanceof Boolean bool) {
      return bool;
    }
    throw new SorafsStorageException("Expected boolean for `" + key + "`");
  }

  private static List<Object> requireList(final Map<String, Object> map, final String key) {
    final Object value = map.get(key);
    if (value instanceof List<?> list) {
      @SuppressWarnings("unchecked")
      final List<Object> cast = (List<Object>) list;
      return cast;
    }
    throw new SorafsStorageException("Expected list for `" + key + "`");
  }

  public String manifestIdHex() {
    return manifestIdHex;
  }

  public String chunkerHandle() {
    return chunkerHandle;
  }

  public String clientId() {
    return clientId;
  }

  public long chunkCount() {
    return chunkCount;
  }

  public long contentLength() {
    return contentLength;
  }

  public long assembledBytes() {
    return assembledBytes;
  }

  public List<ProviderReport> providerReports() {
    return providerReports;
  }

  public List<ChunkReceipt> chunkReceipts() {
    return chunkReceipts;
  }

  public String anonymityPolicy() {
    return anonymityPolicy;
  }

  public String anonymityStatus() {
    return anonymityStatus;
  }

  public String anonymityReason() {
    return anonymityReason;
  }

  public long anonymitySoranetSelected() {
    return anonymitySoranetSelected;
  }

  public long anonymityPqSelected() {
    return anonymityPqSelected;
  }

  public long anonymityClassicalSelected() {
    return anonymityClassicalSelected;
  }

  public double anonymityPqRatio() {
    return anonymityPqRatio;
  }

  public double anonymityClassicalRatio() {
    return anonymityClassicalRatio;
  }

  public double anonymityCandidateRatio() {
    return anonymityCandidateRatio;
  }

  public double anonymityDeficitRatio() {
    return anonymityDeficitRatio;
  }

  public double anonymitySupplyDelta() {
    return anonymitySupplyDelta;
  }

  public boolean anonymityBrownout() {
    return anonymityBrownout;
  }

  public boolean anonymityBrownoutEffective() {
    return anonymityBrownoutEffective;
  }

  public boolean anonymityUsesClassical() {
    return anonymityUsesClassical;
  }

  /** Provider-level outcome summary. */
  public static final class ProviderReport {
    private final String provider;
    private final long successes;
    private final long failures;
    private final boolean disabled;

    private ProviderReport(
        final String provider, final long successes, final long failures, final boolean disabled) {
      this.provider = provider;
      this.successes = successes;
      this.failures = failures;
      this.disabled = disabled;
    }

    public String provider() {
      return provider;
    }

    public long successes() {
      return successes;
    }

    public long failures() {
      return failures;
    }

    public boolean disabled() {
      return disabled;
    }
  }

  /** Chunk-level receipt emitted by the orchestrator. */
  public static final class ChunkReceipt {
    private final int chunkIndex;
    private final String provider;
    private final long attempts;

    private ChunkReceipt(final int chunkIndex, final String provider, final long attempts) {
      this.chunkIndex = chunkIndex;
      this.provider = provider;
      this.attempts = attempts;
    }

    public int chunkIndex() {
      return chunkIndex;
    }

    public String provider() {
      return provider;
    }

    public long attempts() {
      return attempts;
    }
  }

  private static final class Builder {
    private String manifestIdHex;
    private String chunkerHandle;
    private String clientId;
    private long chunkCount;
    private long contentLength;
    private long assembledBytes;
    private String anonymityPolicy;
    private String anonymityStatus;
    private String anonymityReason;
    private long anonymitySoranetSelected;
    private long anonymityPqSelected;
    private long anonymityClassicalSelected;
    private double anonymityClassicalRatio;
    private double anonymityPqRatio;
    private double anonymityCandidateRatio;
    private double anonymityDeficitRatio;
    private double anonymitySupplyDelta;
    private boolean anonymityBrownout;
    private boolean anonymityBrownoutEffective;
    private boolean anonymityUsesClassical;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/sorafs/GatewayFetchSummary.kt
package org.hyperledger.iroha.sdk.sorafs

import org.hyperledger.iroha.sdk.client.JsonParser

/**
 * Immutable view of the JSON summary returned by the SoraFS orchestrator after a gateway fetch.
 */
class GatewayFetchSummary private constructor(
    @JvmField val manifestIdHex: String,
    @JvmField val chunkerHandle: String,
    @JvmField val clientId: String?,
    @JvmField val chunkCount: Long,
    @JvmField val contentLength: Long,
    @JvmField val assembledBytes: Long,
    @JvmField val providerReports: List<ProviderReport>,
    @JvmField val chunkReceipts: List<ChunkReceipt>,
    @JvmField val anonymityPolicy: String,
    @JvmField val anonymityStatus: String,
    @JvmField val anonymityReason: String?,
    @JvmField val anonymitySoranetSelected: Long,
    @JvmField val anonymityPqSelected: Long,
    @JvmField val anonymityClassicalSelected: Long,
    @JvmField val anonymityClassicalRatio: Double,
    @JvmField val anonymityPqRatio: Double,
    @JvmField val anonymityCandidateRatio: Double,
    @JvmField val anonymityDeficitRatio: Double,
    @JvmField val anonymitySupplyDelta: Double,
    @JvmField val anonymityBrownout: Boolean,
    @JvmField val anonymityBrownoutEffective: Boolean,
    @JvmField val anonymityUsesClassical: Boolean,
) {
    /** Provider-level outcome summary. */
    class ProviderReport(
        @JvmField val provider: String,
        @JvmField val successes: Long,
        @JvmField val failures: Long,
        @JvmField val disabled: Boolean,
    )

    /** Chunk-level receipt emitted by the orchestrator. */
    class ChunkReceipt(
        @JvmField val chunkIndex: Int,
        @JvmField val provider: String,
        @JvmField val attempts: Long,
    )

    companion object {
        /** Parses a JSON payload (UTF-8 bytes) into a `GatewayFetchSummary`. */
        @JvmStatic
        fun fromJsonBytes(payload: ByteArray): GatewayFetchSummary {
            require(payload.isNotEmpty()) { "Gateway summary JSON must not be empty" }
            val json = String(payload, Charsets.UTF_8)
            val parsed = JsonParser.parse(json)
            if (parsed !is Map<*, *>) throw SorafsStorageException("Gateway summary must be a JSON object")
            @Suppress("UNCHECKED_CAST")
            return fromJsonObject(parsed as Map<String, Any>)
        }

        @Suppress("UNCHECKED_CAST")
        private fun fromJsonObject(root: Map<String, Any>): GatewayFetchSummary {
            val providerReports = (requireList(root, "provider_reports")).map { item ->
                if (item !is Map<*, *>) throw SorafsStorageException("provider_reports entries must be objects")
                val map = item as Map<String, Any>
                ProviderReport(
                    provider = requireString(map, "provider"),
                    successes = requireLong(map, "successes"),
                    failures = requireLong(map, "failures"),
                    disabled = requireBoolean(map, "disabled"),
                )
            }
            val chunkReceipts = (requireList(root, "chunk_receipts")).map { item ->
                if (item !is Map<*, *>) throw SorafsStorageException("chunk_receipts entries must be objects")
                val map = item as Map<String, Any>
                ChunkReceipt(
                    chunkIndex = requireInt(map, "chunk_index"),
                    provider = requireString(map, "provider"),
                    attempts = requireLong(map, "attempts"),
                )
            }
            return GatewayFetchSummary(
                manifestIdHex = SorafsInputValidator.normalizeHexBytes(
                    requireString(root, "manifest_id_hex"), "manifest_id_hex", 32),
                chunkerHandle = requireString(root, "chunker_handle"),
                clientId = optionalString(root, "client_id"),
                chunkCount = requireLong(root, "chunk_count"),
                contentLength = requireLong(root, "content_length"),
                assembledBytes = requireLong(root, "assembled_bytes"),
                providerReports = providerReports,
                chunkReceipts = chunkReceipts,
                anonymityPolicy = requireString(root, "anonymity_policy"),
                anonymityStatus = requireString(root, "anonymity_status"),
                anonymityReason = optionalString(root, "anonymity_reason"),
                anonymitySoranetSelected = requireLong(root, "anonymity_soranet_selected"),
                anonymityPqSelected = requireLong(root, "anonymity_pq_selected"),
                anonymityClassicalSelected = requireLong(root, "anonymity_classical_selected"),
                anonymityClassicalRatio = requireDouble(root, "anonymity_classical_ratio"),
                anonymityPqRatio = requireDouble(root, "anonymity_pq_ratio"),
                anonymityCandidateRatio = requireDouble(root, "anonymity_candidate_ratio"),
                anonymityDeficitRatio = requireDouble(root, "anonymity_deficit_ratio"),
                anonymitySupplyDelta = requireDouble(root, "anonymity_supply_delta"),
                anonymityBrownout = requireBoolean(root, "anonymity_brownout"),
                anonymityBrownoutEffective = requireBoolean(root, "anonymity_brownout_effective"),
                anonymityUsesClassical = requireBoolean(root, "anonymity_uses_classical"),
            )
        }

        private fun requireString(map: Map<String, Any>, key: String): String {
            val value = map[key]
            if (value is String) return value
            throw SorafsStorageException("Expected string for `$key`")
        }

        private fun optionalString(map: Map<String, Any>, key: String): String? {
            val value = map[key] ?: return null
            if (value is String) return value
            throw SorafsStorageException("Expected string for `$key`")
        }

        private fun requireLong(map: Map<String, Any>, key: String): Long {
            val value = map[key]
            if (value is Number) {
                if (value is Float || value is Double) throw SorafsStorageException("Expected integer for `$key`")
                return value.toLong()
            }
            throw SorafsStorageException("Expected number for `$key`")
        }

        private fun requireInt(map: Map<String, Any>, key: String): Int {
            val value = requireLong(map, key)
            if (value < 0 || value > Int.MAX_VALUE) throw SorafsStorageException("Expected int range for `$key`")
            return value.toInt()
        }

        private fun requireDouble(map: Map<String, Any>, key: String): Double {
            val value = map[key]
            if (value is Number) return value.toDouble()
            throw SorafsStorageException("Expected number for `$key`")
        }

        private fun requireBoolean(map: Map<String, Any>, key: String): Boolean {
            val value = map[key]
            if (value is Boolean) return value
            throw SorafsStorageException("Expected boolean for `$key`")
        }

        @Suppress("UNCHECKED_CAST")
        private fun requireList(map: Map<String, Any>, key: String): List<Any> {
            val value = map[key]
            if (value is List<*>) return value as List<Any>
            throw SorafsStorageException("Expected list for `$key`")
        }
    }
}
